nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -404,6 +404,87 @@ private Type.ClassType typeWithPreservedAnnotations(ParameterizedTypeTree tree)
     return finalType;
   }
 
+  // here we are checking if we should report an error on not considering the @Nullable annotations
+  // for overriden and
+  // overriding method return type should be consistent with the type parameter
+  public boolean DoNullabilityAnnotationsMatch(
+      Symbol.MethodSymbol overriddenMethod,
+      Symbol.MethodSymbol overridingMethod,
+      VisitorState state) {
+
+    // method return type is of type Type variable
+    if (config.isJSpecifyMode()) {
+
+      // type of the type variable in the overridden class
+      // matches the return type of the method
+      Type.MethodType overriddenMethodType =
+          (Type.MethodType)
+              state.getTypes().memberType(overridingMethod.owner.type, overriddenMethod);
+      Type overriddenMethodReturnType =
+          overriddenMethodType.getReturnType(); // this is type of the type parameter
+      // return type of the overriding method
+      Type overridingMethodReturnType = overridingMethod.getReturnType();
+
+      boolean overriddenMethodHasNullableAnnotation =
+          Nullness.hasNullableAnnotation(
+              overriddenMethodReturnType.getAnnotationMirrors().stream(), config);
+
+      boolean overridingMethodHasNullableAnnotation =
+          Nullness.hasNullableAnnotation(
+              overridingMethodReturnType.getAnnotationMirrors().stream(), config);
+
+      if (overriddenMethodHasNullableAnnotation != overridingMethodHasNullableAnnotation) {
+        return false;
+      }
+
+      if (overriddenMethodReturnType instanceof Type.ClassType
+          && overridingMethodReturnType instanceof Type.ClassType) {
+        boolean doNullabilityAnnotationsMatch =
+            compareNullabilityAnnotations(
+                (Type.ClassType) overriddenMethodReturnType,
+                (Type.ClassType) overridingMethodReturnType);
+        if (!doNullabilityAnnotationsMatch) {
+          // don't report an error here as the return type and the type parameters have the same
+          // annotations
+          return false;
+        }
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * if a method has Nullable return type, only then the shouldReportAnError will be called. if the
+   * annotation of R is Nullable and the return types match
+   * node.getTarget().getReceiver().getBlock().getNodes().get(0).rhs.type.tsym member type (type of
+   * f1, sym for apply)
+   */
+  public static Nullness getActualAnnotation(
```

Comment:
This is actually pretty much the exact logic we need for what I wrote above for (1).  Here let's take a `Tree` as the parameter rather than `MethodInvocationNode` so it's more re-usable